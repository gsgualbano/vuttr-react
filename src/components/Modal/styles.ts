import styled from 'styled-components';
import { Theme } from '~/styles/global';

export const Container = styled.div`
  display: flex;
  position: fixed;
  z-index: 1;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  overflow: auto;
  background: rgb(0, 0, 0);
  background: rgba(0, 0, 0, 0.4);
  justify-content: center;
  align-items: center;
`;

export const Content = styled.div`
  color: ${({ theme }: Theme) => theme.colors.textColor};
  display: inline-block;
  background: ${({ theme }: Theme) => theme.colors.modalBackground};
  padding: 10px 20px;
  border: 1px solid;
  border-color: ${({ theme }: Theme) => theme.colors.modalBorderColor};
  box-shadow: 0 0 20px 0 rgba(0, 0, 0, 0.1);
  border-radius: 5px;
  max-width: 540px;
  width: 100%;

  @media (min-width: 768px) {
    padding: 20px 40px;
  }
`;
