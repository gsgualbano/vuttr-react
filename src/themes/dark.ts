export default {
  colors: {
    background: '#1B1A1A',
    textColor: '#fff',
    cardBackground: '#29292B',
    cardBorderColor: '#29292B',
    modalBackground: '#29292B',
    modalBorderColor: '#29292B',
    linkColor: '#ae8fd6',
    linkVisitedColor: '##8050bf',
    buttonOutlineBackground: '#29292B',
    InputBackground: '#9e9fa2',
    InputBorderColor: '#9e9fa2',
  },
};
