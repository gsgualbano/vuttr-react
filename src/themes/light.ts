export default {
  colors: {
    background: '#fcfcfd',
    textColor: '#170c3a',
    cardBackground: '#fcfcfd',
    cardBorderColor: '#ebeaed',
    modalBackground: '#fefefe',
    modalBorderColor: '#888',
    linkColor: 'rgb(0, 0, 238)',
    linkVisitedColor: 'rgb(85, 26, 139);',
    buttonOutlineBackground: '#ffffff',
    InputBackground: '#f5f4f6',
    InputBorderColor: '#ebeaed',
  },
};
