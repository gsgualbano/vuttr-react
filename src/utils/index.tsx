/* eslint-disable max-len */
import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import { render } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';
import { reducer } from '../store/ducks/tools';

interface Theme {
  colors: {
    background: string;
    textColor: string;
    cardBackground: string;
    cardBorderColor: string;
    modalBackground: string;
    modalBorderColor: string;
    linkColor: string;
    linkVisitedColor: string;
    buttonOutlineBackground: string;
    InputBackground: string;
    InputBorderColor: string;
  };
}

/** redux mock to render components that depends on the redux */
export const renderWithRedux = (
  component: JSX.Element,
  { initialState, store = createStore(reducer, initialState) }: any = {},
) => ({
  ...render(<Provider store={store}>{component}</Provider>),
  store,
});

/** render components thats depends on the theme provider */
export const renderWithTheme = (component: JSX.Element, theme: Theme) => render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);

/** render components thats depends on the redux and theme provider */
export const renderWithReduxAndTheme = (
  component: JSX.Element,
  theme: Theme,
  { initialState, store = createStore(reducer, initialState) }: any = {},
) => ({
  ...render(
    <ThemeProvider theme={theme}>
      <Provider store={store}>{component}</Provider>
    </ThemeProvider>,
  ),
  store,
});
