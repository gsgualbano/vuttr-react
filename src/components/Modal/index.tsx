import React from 'react';

import { Animated } from 'react-animated-css';
import { Container, Content } from './styles';

/**
 * Props do Modal
 */
interface OwnProps {
  /** se o modal está visível */
  visible?: boolean;
  /** dispatcher para mostrar o modal */
  setShowModal: React.Dispatch<React.SetStateAction<boolean>>;
  /** elemento jsx  */
  children?: React.ReactNode;
}

/**
 * Modal.
 */
const Modal = ({ visible = false, setShowModal, children }: OwnProps): JSX.Element => (
  <>
    {visible && (
      <Animated isVisible={visible} animationIn="fadeIn" animationOut="fadeOut">
        <Container onClick={() => setShowModal(false)} data-testid="outside-modal">
          <Content onClick={e => e.stopPropagation()} data-testid="inside-modal">
            {children}
          </Content>
        </Container>
      </Animated>
    )}
  </>
);

export default Modal;
