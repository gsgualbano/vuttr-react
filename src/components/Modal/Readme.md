Modal examplo:

```js
import { useState } from "react";

const [showModal, setShowModal] = useState(false);

const handleOpen = () => {
  setShowModal(true);
};

const handleClose = () => {
  setShowModal(false);
};

<>
  <button onClick={handleOpen}>Open Modal!</button>
  <Modal visible={showModal} setShowModal={setShowModal}>
    <p>Example Modal!</p>
    <div>
      <button onClick={handleClose}>Close Modal!</button>
    </div>
  </Modal>
</>;
```

Dark Modal exemplo

```js
import { useState } from "react";
import { ThemeProvider } from "styled-components";
import dark from "~/themes/dark";

const [showModal, setShowModal] = useState(false);

const handleOpen = () => {
  setShowModal(true);
};

const handleClose = () => {
  setShowModal(false);
};

<>
  <button onClick={handleOpen}>Open Modal!</button>
  <ThemeProvider theme={dark}>
    <Modal visible={showModal} setShowModal={setShowModal}>
      <p>Example Modal!</p>
      <div>
        <button onClick={handleClose}>Close Modal!</button>
      </div>
    </Modal>
  </ThemeProvider>
</>;
```
