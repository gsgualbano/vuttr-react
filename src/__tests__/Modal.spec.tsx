import React from 'react';
import { renderWithReduxAndTheme } from '~/utils';
import { cleanup } from '@testing-library/react';

import Modal from '~/components/Modal';

import 'jest-dom/extend-expect';
import 'jest-styled-components';

import dark from '~/themes/dark';
import light from '~/themes/light';

describe('<Modal />', () => {
  afterEach(cleanup);

  const setShowModal = () => {};

  it('render correctly', () => {
    const { asFragment } = renderWithReduxAndTheme(
      <Modal visible setShowModal={setShowModal} />,
      light,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('not show when not visible', () => {
    const { queryByTestId } = renderWithReduxAndTheme(<Modal setShowModal={setShowModal} />, light);
    expect(queryByTestId('outside-modal')).not.toBeInTheDocument();
  });

  it('show when visible', () => {
    const { queryByTestId } = renderWithReduxAndTheme(
      <Modal visible setShowModal={setShowModal} />,
      light,
    );
    expect(queryByTestId('outside-modal')).toBeInTheDocument();
  });

  it('render with light theme', () => {
    const { getByTestId } = renderWithReduxAndTheme(
      <Modal visible setShowModal={setShowModal} />,
      light,
    );
    expect(getByTestId('inside-modal')).toHaveStyleRule('background', '#fefefe');
    expect(getByTestId('inside-modal')).toHaveStyleRule('color', '#170c3a');
    expect(getByTestId('inside-modal')).toHaveStyleRule('border-color', '#888');
  });

  it('render with dark theme', () => {
    const { getByTestId } = renderWithReduxAndTheme(
      <Modal visible setShowModal={setShowModal} />,
      dark,
    );
    expect(getByTestId('inside-modal')).toHaveStyleRule('background', '#29292B');
    expect(getByTestId('inside-modal')).toHaveStyleRule('color', '#fff');
    expect(getByTestId('inside-modal')).toHaveStyleRule('border-color', '#29292B');
  });
});
