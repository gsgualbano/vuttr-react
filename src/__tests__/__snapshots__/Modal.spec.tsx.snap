// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> render correctly 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  z-index: 1;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  overflow: auto;
  background: rgb(0,0,0);
  background: rgba(0,0,0,0.4);
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  color: #170c3a;
  display: inline-block;
  background: #fefefe;
  padding: 10px 20px;
  border: 1px solid;
  border-color: #888;
  box-shadow: 0 0 20px 0 rgba(0,0,0,0.1);
  border-radius: 5px;
  max-width: 540px;
  width: 100%;
}

@media (min-width:768px) {
  .c1 {
    padding: 20px 40px;
  }
}

<div
    class="animated fadeIn"
    style="animation-delay: 0ms; animation-duration: 1000ms; pointer-events: all;"
  >
    <div
      class="c0"
      data-testid="outside-modal"
    >
      <div
        class="c1"
        data-testid="inside-modal"
      />
    </div>
  </div>
</DocumentFragment>
`;
